annotation-server/src/guidelines/guidelines.service.ts

```diff
@@ -0,0 +1,322 @@
+                ),
+            ],
+        );
+
+        this.spreadsheetPhenotypeHeader = phenotypeHeader[0].map(
+            (cell) =>
+                new Set(
+                    cell.value
+                        .split(';')
+                        .map((phenotype) => phenotype.trim().toLowerCase()),
+                ),
+        );
+        for (let row = 0; row < medications.length; row++) {
+            const geneSymbolName = genes[row]?.[0];
+            const medicationName = medications[row]?.[0];
+            if (!geneSymbolName || !medicationName) continue;
+
+            const medication = await this.findMedicationByName(
+                medicationName.value,
+            );
+            const genePhenotypes = await this.findGenePhenotypes(
+                geneSymbolName.value,
+            );
+            if (genePhenotypes.length === 0 || !medication) continue;
+
+            for (let col = 0; col < implications[row].length; col++) {
+                for (const genePhenotype of genePhenotypes[col].values()) {
+                    const guidelinesForMedication = guidelines.get(
+                        medication.name,
+                    );
+                    const guidelinesForGenePhenotype =
+                        guidelinesForMedication.filter(
+                            (guidelineForMed) =>
+                                guidelineForMed.genePhenotype.id ===
+                                genePhenotype.id,
+                        );
+                    if (!guidelinesForGenePhenotype.length)
+                        throw new InternalServerErrorException(
+                            `No matching CPIC guideline was found for ${medication.name} and genephenotype ${genePhenotype.geneSymbol}, ${genePhenotype.phenotype.name}!`,
+                        );
+                    for (const guideline of guidelinesForGenePhenotype) {
+                        const implication =
+                            implications[row][col].value?.trim();
+                        const recommendation =
+                            recommendations[row][col].value?.trim();
+                        const warningLevel = this.getWarningLevelFromColor(
+                            recommendations[row][col].backgroundColor,
+                        );
+                        if (
+                            !implication ||
+                            implication.replace(' ', '').toLowerCase() ===
+                                'n/a' ||
+                            !recommendation ||
+                            recommendation.replace(' ', '').toLowerCase() ===
+                                'n/a'
+                        ) {
+                            continue;
+                        }
+                        guideline.implication = implication;
+                        guideline.recommendation = recommendation;
+                        guideline.warningLevel = warningLevel;
+                    }
```

Comment:
The encapsulating method is already very long and this is the fourth nested `for`-loop. I think we should split up the method a bit; this inner-most `for`-loop specifically seems to be easy to make into it's own method for example.